### PR TITLE
Implement feature value positions to core and back office

### LINKED ISSUE #35499: Add possibility to change position of feature values
**US**
As a merchant, I want to be able to sort the feature position to follow a display logic of my product feature. 

**Specs**
- Add position column on the feature value page. New db filed
- Add endpoint to update position 

- Drag and drop row like on feature group page
Sort by position is selected by default

- Upgrade PR to do

- To adadpt the theme classic and HB, the features from the same feature group follows the position orders and place at the end the custom value

**Acceptance test**
Go in Shop Parameters > Product Settings
Change `Order of feature values` and select `position` instead of `name`
Go on composition 
Drag Polyester below Wool
Check FO Wool is before Polyester

cf https://github.com/PrestaShop/PrestaShop/discussions/32267

### PR DESCRIPTION
**Pull request by ks129**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add a new field to feature value table/model - position. Add it to feature value grid and implement route to change positions.
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Check the linked issue.
| UI Tests          | https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/11725048109
| Fixed issue or discussion?     | Fixes #35499 , Fixes https://github.com/PrestaShop/PrestaShop/discussions/32267
| Related PRs       | PrestaShop/autoupgrade#925 , PrestaShop/ps_facetedsearch#1083
| Sponsor company   | -

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `category` should be one of these: `FO`, `BO`, `CO`, `IN`, `WS`, `TE`, `LO`, `ME` or `PM`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

File: src/Core/Grid/Definition/Factory/FeatureValueGridDefinitionFactory.php
Comment (jolelievre) [inline src/Core/Grid/Definition/Factory/FeatureValueGridDefinitionFactory.php]:
```suggestion
            ->add((new Filter('position', ReorderPositionsButtonType::class))
                ->setAssociatedColumn('position')
            )
```

Comment (jolelievre) [reply src/Core/Grid/Definition/Factory/FeatureValueGridDefinitionFactory.php]: We made some changes in all the grids that handle position, we removed the filter by position (which is not very useful) and instead we use this dedicated type that helps to set up the grid in a "positionable state"

A grid can only be ordered when it is sorted by position, and this custom type adds a button that helps the user understand this.

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureValueController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureValueController.php:335]: You don't seem shaken at all by the new `PrestaShopAdminController` even without the doc being updated yet, great job!

**Review by jolelievre — COMMENTED** [review]
Thanks you @ks129 

That's a perfect contribution, you anticipated all the side changes. Great job!

I just had one suggestion in the grid definition, but you couldn't be aware of this new UX we want to uniformize on all the positionable grids.

I also have a small request, if you can add it in this PR it would be perfect, if you prefer to handle it separately it's understandable.

The list of features in the product page should also reflect this change, from what I found it seems to be handled in the [ProductLazyArray](https://github.com/PrestaShop/PrestaShop/blob/892b08345199708351fe5854ee855ddd1fff9864/src/Adapter/Presenter/Product/ProductLazyArray.php#L421) or in the [Product ObjectModel](https://github.com/PrestaShop/PrestaShop/blob/564dbcc461aabf2165da30ad8270a1101ae499a5/classes/Product.php#L5715), the idea would be to be able to sort feature values by position as well, and the tip top feature would be to have a configuration to choose between Alphabetical or Position order like you did in faceted search.

In a nutshell it would mean:
- add a configuration value in Product settings > Product page form
- adapt the order by in FO depending on this configuration

Would you accept to add this in the PR please?

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Shopparameters.Feature`
    - [ ] `Order of feature values`
- `Admin.Shopparameters.Help`
    - [ ] `In the product page, feature values can be sorted alphabetically by value or by position.`

(Note: this is an automated message, but answering it will reach a real human)

**Comment by gericfo** [comment]
Hello @ks129 
Please note that the **“Needs autoupgrade PR”** label has been automatically added to your Pull Request. This means that the changes made in your PR must also be reported in the **Autoupgrade module**, responsible for running the update process of PrestaShop stores. 
When there's a modification to the `db-structure.sql` file or if a Doctrine entity has been modified, added, or deleted, it impacts the database structure. Modifying the list of hooks and feature flags has consequences on the update process as well. These changes can affect how the application interacts with the database and how update procedures are handled. It's important to consider these aspects when contributing to ensure a smooth update experience for users. 
That's why this contribution should be backported, with another corresponding PR, on the **Autoupgrade module's GitHub repository** which is available here: [link](https://github.com/PrestaShop/autoupgrade). 
Don't forget to add a comment in this current thread with the link to this PR once it has been created. 
Thank you.

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: Can you use string values instead please? so we understand the meaning of each config
If you prefer to keep the int value it's fine but then we'd need some const matching these values for better understanding (which is why I suggested the string instead: `name`, `position`)

**Review by jolelievre — COMMENTED** [review]
Thanks @ lot @ks129 

Only one remaining comment, and the I'll push for this to go quickly to QA so that it's included in v9

**Review by kpodemski — CHANGES_REQUESTED** [review]
Hello

We recently introduced the 9.0.x branch, which will be the base for PrestaShop 9.0. If you want to see your change in this version, please rebase your PR onto 9.0.x. The `develop` will be used for PrestaShop 9.1, so you need to decide what to do.

Please rebase onto `9.0.x` or let me know your decision if you don't need to have it there so I can unblock the pull request.

**Comment by ks129** [comment]
@kpodemski @jolelievre Both things are done now. However, I have a little problem with UI tests as I don't understand why they aren't working.

**Review by jolelievre — DISMISSED** [review]

**Comment by jolelievre** [comment]
@ks129 we have issues with the test workflow you used, it's planned to be fixed but I don't have time for it right now

You can use this workflow instead https://github.com/ks129/ga.tests.ui.pr/actions/workflows/pr_test_one.yml which is not buggy in theory

**Review by jolelievre — DISMISSED** [review]

**Comment by jolelievre** [comment]
Hi @ks129 

I allowed myself to rebase your PR, and made some small modifications in UI tests to match some recent changes in them.

**Review by Progi1984 — APPROVED** [review]
Ok for UI Tests

**Review by kpodemski — DISMISSED** [review]

**Comment by kpodemski** [comment]
well done @ks129 👏🏻

**Review by florine2623 — COMMENTED** [review]
Hello @ks129 ,

Regarding the [specs](https://github.com/PrestaShop/PrestaShop/issues/35499), I checked the following :
- Add position column on the feature value page. New db filed. ✅ 
- Drag and drop row like on feature group page
Sort by position is selected by default ✅ 

Drag and drop from BO works as expected. Database is filed with position in feature_value table.

**In BO :**
Screenshot 2024-11-06 at 16 49 57

**In DB :**
Screenshot 2024-11-06 at 16 48 40

Button Rearrange will rearrange the Position if the grid is sorted by ID or value. Position can be sorted with the arrow button too.

Nothing is changed on FO as this PR only tacles BO part..
The acceptance test is not completed for the FO section only the BO section.

> Acceptance test
> Go on composition
> Drag Polyester below Wool
> Check FO Wool is before Polyester

I can't validate it as automated tests are broken : https://github.com/florine2623/testing_pr/actions/runs/11706411660/job/32607716941#step:10:309 :x:

@PrestaShop/qa-automation & @ks129 could you check please ?

Thanks!

File: tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts]:
```suggestion
    const pageTitle = await boAttributesPage.getPageTitle(page);
    expect(pageTitle).to.contains(boAttributesPage.pageTitle);
```

File: tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts]:
```suggestion
    await boAttributesPage.goToFeaturesPage(page);
```

File: tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/04_attributesAndFeatures/02_features/02_changePosition.ts]:
```suggestion
  boAttributesPage,
  boDashboardPage,
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@ks129 Small changes in UI Tests after move to the ui testling library

**Comment by Progi1984** [comment]
@ks129 I have added some changes for validating UI Tests

**Review by jolelievre — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]
OK for UI Tests

**Comment by jolelievre** [comment]
@florine2623 I just update the Acceptance test, actually for the position to be used in FO you need to update a configuration in Shop Parameters > Product Settings > Order of feature values => select position

**Review by tleon — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello all,

Tested again this feature with correct configuration of `Order of feature values`. It works as expected in the FO.

The automated tests are green : https://github.com/florine2623/testing_pr/actions/runs/11801417927

Thanks! It is QA ✅

**Pull request merged**
